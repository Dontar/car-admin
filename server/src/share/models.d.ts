export interface ICars {
    id: number;
    rama: string;
    dkn: string;
    mark_name: string;
    model_name: string;
    produce_year: string;
    client_id?: any;
}

export interface ICompany {
    id: number;
    client_name: string;
    bulstat: string;
    representative: string;
    phone: string;
    mobile: string;
    region_name: string;
    municipality_name: string;
    city: string;
    postcode: string;
    street_name: string;
    street_no: string;
    blok: string;
    vhod: string;
    apartment: string;
    floor: string;
    vehicle_ids: string;
}

export interface IPerson {
    id: number;
    client_name: string;
    egn: string;
    representative: string;
    phone: string;
    mobile: string;
    region_name: string;
    municipality_name: string;
    city: string;
    postcode: string;
    street_name: string;
    street_no: string;
    blok: string;
    vhod: string;
    apartment: string;
    floor: string;
    vehicle_ids: string;
}

type Identifier = string | number;
interface RecordType {
    id: Identifier;
}

export type GetParams = GetListParams & GetManyParams & GetManyReferenceParams;
export type GetResult<T> = Partial<GetListResult<T> & GetManyResult<T> & GetManyReferenceResult<T>>;

export interface GetListParams {
    pagination: {
        page: number;
        perPage: number;
    };
    sort: {
        field: string;
        order: string;
    };
    filter: any;
}

export interface GetListResult<T> {
    data: (RecordType & T)[];
    total: number;
    validUntil?: Date;
}
export interface GetOneParams {
    id: Identifier;
}
export interface GetOneResult<T> {
    data: RecordType & T;
    validUntil?: Date;
}
export interface GetManyParams {
    ids: Identifier[];
}
export interface GetManyResult<T> {
    data: (RecordType & T)[];
    validUntil?: Date;
}
export interface GetManyReferenceParams extends GetListParams{
    target: string;
    id: Identifier;
}
export interface GetManyReferenceResult<T> {
    data: (RecordType & T)[];
    total: number;
    validUntil?: Date;
}
export interface UpdateParams<T = any> {
    id: Identifier;
    data: T;
    previousData: RecordType;
}
export interface UpdateResult {
    data: RecordType;
    validUntil?: Date;
}
export interface UpdateManyParams<T = any> {
    ids: Identifier[];
    data: T;
}
export interface UpdateManyResult {
    data?: Identifier[];
    validUntil?: Date;
}
export interface CreateParams<T = any> {
    data: T;
}
export interface CreateResult {
    data: RecordType;
    validUntil?: Date;
}
export interface DeleteParams {
    id: Identifier;
    previousData: Record;
}
export interface DeleteResult {
    data?: RecordType;
}
export interface DeleteManyParams {
    ids: Identifier[];
}
export interface DeleteManyResult {
    data?: Identifier[];
}
