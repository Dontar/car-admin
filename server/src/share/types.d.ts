declare module 'sql-bricks-sqlite' {
    import sql, { Statement } from 'sql-bricks'

    export interface SelectStatement extends Statement {
        limit(limit: number): SelectStatement;
        offset(offset: number): SelectStatement;
        // orReplace()
        // orAbort()
    }

    export = sql;
}