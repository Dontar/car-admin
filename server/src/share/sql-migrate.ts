import { readdir, readFile } from 'fs/promises';
import * as path from 'path';
import { MyDatabase } from '../connection';
import { migrate } from './models';

type MigrationFile = migrate.MigrationFile
type MigrationParams = migrate.MigrationParams
type MigrationData = migrate.MigrationData

export async function readMigrations(migrationPath?: string): Promise<Array<MigrationData>> {
    const migrationsPath = migrationPath || path.join(process.cwd(), 'migrations');
    const location = path.resolve(migrationsPath);

    const migrationFiles = (await readdir(location)).reduce((a, file) => {
        const [filename, id, name] = file.match(/^(\d+).(.*?)\.sql$/) ?? [];
        if (filename) {
            a.push({ id: Number(id), name, filename });
        }
        return a;
    }, [] as MigrationFile[]);

    if (!migrationFiles.length) {
        throw new Error(`No migration files found in '${location}'.`);
    }

    // Get the list of migrations, for example:
    //   { id: 1, name: 'initial', filename: '001-initial.sql', up: ..., down: ... }
    //   { id: 2, name: 'feature', filename: '002-feature.sql', up: ..., down: ... }
    return Promise.all(
        migrationFiles.map(migration =>
            readFile(path.join(location, migration.filename), 'utf8').then(data => {
                const [up, down] = data.split(/^--\s+?down\b/im);
                return {
                    ...migration,
                    up: up.replace(/^-- .*?$/gm, '').trim(),
                    down: down ? down.trim() : ''
                };
            })));
}

/**
 * Migrates database schema to the latest version
 */
export async function migrate(db: MyDatabase, config: MigrationParams = {}): Promise<void> {
    config.force = config.force || false;
    config.table = config.table || 'migrations';

    const { force, table } = config;
    const migrations = config.migrations
        ? config.migrations
        : await readMigrations(config.migrationsPath);

    // Create a database table for migrations meta data if it doesn't exist
    db.prepare(`
        CREATE TABLE IF NOT EXISTS "${table}" (
            id   INTEGER PRIMARY KEY,
            name TEXT    NOT NULL,
            up   BLOB    NOT NULL,
            down BLOB    NOT NULL
        )
    `).run();

    // Get the list of already applied migrations
    let dbMigrations: migrate.MigrationData[] = db.prepare(
        `SELECT id, name, up, down FROM "${table}" ORDER BY id ASC`
    ).all();

    const deleteMigration = db.prepare(`DELETE FROM "${table}" WHERE id = ?`);
    const undoMigration = db.transaction((migration) => {
        db.exec(migration.down);
        deleteMigration.run(migration.id);
        dbMigrations = dbMigrations.filter(x => x.id !== migration.id);
    });

    // Undo migrations that exist only in the database but not in files,
    // also undo the last migration if the `force` option is enabled.
    const lastMigration = migrations[migrations.length - 1];
    for (const migration of dbMigrations.slice().reverse()) {
        if (existsOnlyInDb(migration)) {
            undoMigration(migration);
        } else {
            break;
        }
    }

    const insertMigration = db.prepare(`INSERT INTO "${table}" (id, name, up, down) VALUES (?, ?, ?, ?)`);
    const addMigration = db.transaction((migration: migrate.MigrationData) => {
        db.exec(migration.up);
        insertMigration.run(
            migration.id,
            migration.name,
            migration.up,
            migration.down
        );
    });

    // Apply pending migrations
    const lastMigrationId = dbMigrations.length
        ? dbMigrations[dbMigrations.length - 1].id
        : 0;
    for (const migration of migrations) {
        if (migration.id > lastMigrationId) {
            addMigration(migration);
        }
    }

    function existsOnlyInDb(migration: migrate.MigrationData) {
        return !migrations.some(x => x.id === migration.id) ||
            (force && migration.id === lastMigration.id);
    }
}
