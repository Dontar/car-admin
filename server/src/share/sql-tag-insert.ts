import { ISql, sql } from './sql-tag';

export function insertSql(table: string, fields: Record<string, unknown>): ISql {
    const values: unknown[] = [];
    const fieldsToString = Object.entries(fields).reduce((a, [key, value], idx) => {
        a += idx > 0 ? ', ' + key : key;
        values.push(value);
        return a;
    }, '');

    const result: unknown = [
        `insert into ${table} (${fieldsToString}) values (`,
        ...Array.from(', '.repeat(values.length - 1)),
        ')'
    ] as const;
    return sql(result as TemplateStringsArray, ...values);
}
