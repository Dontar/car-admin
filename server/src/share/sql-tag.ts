import { createSqlStream, SqlStream } from './SqlStream';
import { createHash } from 'crypto';
import { insertSql } from './sql-tag-insert';

/************
 * Interface
 ************/

export interface ISql<T = unknown> {
    readonly text: string;
    readonly values: Array<unknown>;
    readonly stream: SqlStream<T>;
    readonly where: IWhere | undefined;
    readonly hash: string;
    append(qry: ISql<unknown> | IWhere | string): void;

    readonly type: 'sql'
}

export type WhereParams =
    | [object: Record<string, unknown>]
    | [field: string | ISql | IWhere]
    | [field: string | ISql | IWhere, value: unknown];

export interface IWhere {
    readonly text: string;
    readonly values: unknown[];
    addWhere: boolean;

    and(...args: WhereParams): this;
    or(...args: WhereParams): this;

    readonly type: 'where';
}

export function sql<T = unknown>(str: TemplateStringsArray, ...args: unknown[]): ISql<T> {
    return new Sql(str, ...args);
}

export function where(...params: WhereParams | []): IWhere {
    return new Where(...params);
}

export function tag(...params: WhereParams | []): IWhere {
    const where = new Where(...params);
    where.addWhere = false;
    return where;
}

export function isSql(obj: unknown): obj is ISql {
    return obj !== undefined && (obj as ISql).type === 'sql';
}

export function isWhere(obj: unknown): obj is IWhere {
    return obj !== undefined && (obj as IWhere).type === 'where';
}

export function isSqlOrWhere(obj: unknown): obj is ISql | IWhere {
    return isSql(obj) || isWhere(obj);
}

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace sql {
    export const insert = insertSql;
}

/*************
 * Internals
 *************/

class Where implements IWhere {

    public type: 'where' = 'where';
    public addWhere = true;
    public get text() {
        const result: string[] = [];
        for (const [idx, [op, field, val]] of this.whereExpressions.entries()) {
            let expression = idx > 0 ? `${op} ` : '';
            expression += `(${field}`;
            if (val !== undefined) {
                if (isSqlOrWhere(val)) {
                    if (isWhere(val)) {
                        val.addWhere = false;
                    }
                    expression += val.text;
                } else {
                    expression += Array.isArray(val) ? `(${'?, '.repeat(val.length - 1)}?)` : '?';
                }
            }
            expression += ')';
            result.push(expression);
        }

        if (this.addWhere && result.length > 0) {
            result.unshift('WHERE');
        }

        return result.join(' ');
    }
    public get values() {
        return this.whereExpressions.flatMap(([, , val]) => {
            if (isSqlOrWhere(val)) {
                return val.values;
            }
            return val;

        }).filter(i => i !== undefined);
    }

    public and = this._add.bind(this, 'and');
    public or = this._add.bind(this, 'or');

    private whereExpressions: [op: 'and' | 'or', field: string | ISql | IWhere, val: unknown][] = [];

    constructor(...params: WhereParams | []) {
        if (params.length) {
            this._add('and', ...params);
        }
    }

    private _add(op: 'and' | 'or', ...args: WhereParams) {
        const [field, value] = args;
        if (typeof field == 'string' || isSqlOrWhere(field)) {
            this.whereExpressions.push([op, field, value]);
        } else {
            this.whereExpressions = this.whereExpressions.concat(Object.entries(field).map(e => [op, ...e]));
        }
        return this;
    }
}

class Sql<T = unknown> implements ISql<T> {
    type: 'sql' = 'sql';

    private additionalSql: (ISql<unknown> | IWhere | string)[] = [];

    private args: unknown[];
    constructor(private str: TemplateStringsArray, ...args: unknown[]) {
        this.args = args;
    }

    get text() {
        let result = '';
        for (const [idx, val] of this.args.entries()) {
            result += this.str[idx];
            if (isSqlOrWhere(val)) {
                result += val.text;
            } else if (Array.isArray(val)) {
                result += `(${'?, '.repeat(val.length - 1)}?)`;
            } else {
                result += '?';
            }
        }
        result += this.str[this.str.length - 1];
        for (const val of this.additionalSql) {
            if (isSqlOrWhere(val)) {
                result += '\n' + val.text;
            } else {
                result += '\n' + val;
            }
        }
        return result;
    }
    get values() {
        return this.args
            .flatMap(val => isSqlOrWhere(val) ? val.values : val)
            .concat(this.additionalSql.flatMap(i => isSqlOrWhere(i) ? i.values : undefined))
            .filter(i => i !== undefined);
    }
    get stream(): SqlStream<T> {
        return createSqlStream(this) as SqlStream<T>;
    }
    get where(): IWhere | undefined {
        return [...this.args, ...this.additionalSql].find(i => isWhere(i)) as IWhere | undefined;
    }
    get hash() {
        return createHash('sha256').update(JSON.stringify(this), 'utf8').digest('hex');
    }
    append(qry: ISql<unknown> | IWhere | string) {
        this.additionalSql.push(qry);
    }
    toJSON() {
        const { text, values } = this;
        return { text, values };
    }
}
