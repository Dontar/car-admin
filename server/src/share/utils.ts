import { getDB } from '../connection';
import { GetManyReferenceParams, GetOneParams, GetParams } from './models';
import { ISql, sql } from './sql-tag';

export function isOne(params: unknown): params is GetOneParams {
    function isDefined(val: unknown) {
        return val !== null && val !== undefined;
    }
    return isDefined((<GetOneParams>params).id) && !isDefined((<GetManyReferenceParams>params).target);
}

export function getSqlCount(qry: ISql): number {
    const { text, values, hash } = sql`select COUNT(*) as total from(${qry})`;
    let total = 0;
    return getSqlCount.countCache.get(hash) ??
        getSqlCount.countCache.set(hash, total = (getDB().prepare(text).get(values) ?? { total: 0 }).total), total;
}

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace getSqlCount {
    export const countCache = new Map<string, number>();
}

export function processGetManyRefs(params: Partial<GetParams>, qry: ISql): boolean {
    if (params.target) {
        qry.where?.and(params.target + ' = ', params.id);
        return true;
    }
    return false;
}

export function processPaginationAndSort(params: Partial<GetParams>, qry: ISql): {
    sort: boolean;
    pagination: boolean;
} {
    // let total: number | undefined = undefined;
    const result = { sort: false, pagination: false };
    if (params.sort) {
        const { field, order } = params.sort;
        qry.append(`order by ${field} ${order}`);
        result.sort = true;
    }

    if (params.pagination) {
        const { page, perPage } = params.pagination;
        qry.append(sql`limit ${perPage}`);
        qry.append(sql`offset ${(page - 1) * perPage}`);
        result.pagination = true;
    }
    return result;
}

export function stringIdsToArray<T>(obj: T): T;
export function stringIdsToArray<T>(obj: T | T[]): T | T[] {
    type Obj = Record<string, unknown>;

    function transform(obj: Obj) {
        for (const field of Object.keys(obj).filter(field => field.substr(- 4) === '_ids')) {
            const val = obj[field];
            if (typeof val === 'string' && /^\{|\[/.test(val[0])) {
                try {
                    obj[field] = val != '[null]' ? JSON.parse(val) : [];
                } catch (e) {
                    obj[field] = val;
                }
            }
        }
        return obj;
    }

    if (Array.isArray(obj)) {
        return obj.map(obj => transform(obj as Obj) as T);
    } else {
        return transform(obj as Obj) as T;
    }
}
