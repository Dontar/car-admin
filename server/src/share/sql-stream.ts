import { getDB } from '../connection';
import { Readable } from 'stream';
import { ISql, isSql } from './sql-tag';
import { stringIdsToArray } from './utils';

export type SqlStream<T> = NodeJS.ReadableStream & AsyncIterable<T>;

export function createSqlStream<T>(qry: string | ISql, ...params: unknown[]): SqlStream<T> {
    let dataSource: IterableIterator<T>;
    return new Readable({
        objectMode: true,
        read() {
            try {
                if (!dataSource) {
                    const db = getDB();
                    const stmt = db.prepare(isSql(qry) ? qry.text : qry);
                    const iterator = stmt.iterate(...(isSql(qry) ? qry.values : params));
                    dataSource = iterator;
                }
                const { value, done } = dataSource.next();
                this.push(!done ? stringIdsToArray(value) : null);
            } catch (e) {
                this.destroy(e);
            }
        }
    });
}
