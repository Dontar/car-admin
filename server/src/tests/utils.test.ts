import { config } from 'dotenv';
import { IPerson } from '../share/models';
import { createSqlStream } from '../share/SqlStream';
config();

jest.setTimeout(5000 * 10);

describe('Test Sql & Json Streams', () => {

    it('Sql Stream returns correct result', async () => {
        const sql = 'SELECT * FROM CLIENTS LIMIT 5';
        const stream = createSqlStream<IPerson>(sql);
        const result = [];
        for await (const row of stream) {
            result.push(row);
        }
        expect(result).toMatchSnapshot();
    });

});
