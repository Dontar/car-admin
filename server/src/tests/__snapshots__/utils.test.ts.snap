// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Sql & Json Streams Sql Stream returns correct result 1`] = `
Array [
  Object {
    "apartment": 0,
    "blok": "0",
    "city": "РУСЕ",
    "client_name": "Юрий Николов Величков",
    "client_type": 1,
    "driving_practice": 5,
    "egn": "6104266565",
    "floor": 1,
    "id": 9213,
    "minucipality_name": "Русе",
    "mobile": "0889827210",
    "phone": "9294130",
    "postcode": 7000,
    "region_name": "РУСЕ",
    "representative": "",
    "street_name": "СЛАВЯНСКИ",
    "street_no": "14",
    "vhod": "1",
  },
  Object {
    "apartment": 2,
    "blok": "18",
    "city": "СИЛИСТРА",
    "client_name": "First Middle Last",
    "client_type": 1,
    "driving_practice": 5,
    "egn": "6310305565",
    "floor": 1,
    "id": 9219,
    "minucipality_name": "Силистра",
    "mobile": "0891043493",
    "phone": "02830123",
    "postcode": 7500,
    "region_name": "СИЛИСТРА",
    "representative": "",
    "street_name": "ГЕНЕРАЛ  СКОБЕЛЕВ",
    "street_no": "",
    "vhod": "А",
  },
  Object {
    "apartment": 0,
    "blok": "0",
    "city": "СОФИЯ",
    "client_name": "СИ ЕФ ЕМ ООД  ",
    "client_type": 2,
    "driving_practice": 10,
    "egn": "131304899",
    "floor": 0,
    "id": 9251,
    "minucipality_name": "Столична",
    "mobile": "359889100001",
    "phone": "359889100001",
    "postcode": 1000,
    "region_name": "София (столица)",
    "representative": "1",
    "street_name": "УЛ. НИКОЛАЙ ХАЙТОВ БЛ. 2, ВХ. Д, ОФИСИ 8 И 9",
    "street_no": "0",
    "vhod": "0",
  },
  Object {
    "apartment": 0,
    "blok": "0",
    "city": "ЛОЗЕН",
    "client_name": "Бонев Транс ЕООД  ",
    "client_type": 2,
    "driving_practice": 0,
    "egn": "175153616",
    "floor": 0,
    "id": 9351,
    "minucipality_name": "Столична",
    "mobile": "0",
    "phone": "0",
    "postcode": 1151,
    "region_name": "София (столица)",
    "representative": "Генади Стефанов Бонев",
    "street_name": "ЯНКО ПАНАЙОТОВ",
    "street_no": "28",
    "vhod": "0",
  },
  Object {
    "apartment": 7,
    "blok": "9",
    "city": "СОФИЯ",
    "client_name": "Николинка Стефанова Цолова",
    "client_type": 1,
    "driving_practice": 5,
    "egn": "5507181593",
    "floor": 4,
    "id": 9451,
    "minucipality_name": "Столична",
    "mobile": "0897527002",
    "phone": "0888998899",
    "postcode": 1000,
    "region_name": "София (столица)",
    "representative": "",
    "street_name": "ЛЮБЛАНА",
    "street_no": "1",
    "vhod": "a",
  },
]
`;
